### PR TITLE
💡 Proposing for SNlM0e data in `queryBard` and `askAI` functions parameter and `class Chat` property

### PR DESCRIPTION
**Pull request by Aldhanekaa**
# Why this can be great feature for the library?
As the code base written in [Index.js](https://github.com/EvanZhouDev/bard-ai/blob/main/index.js):
- In `init` function, it returns `SNlM0e` data. However, the return of this data can't make impact to the library. Because none of the function ([`queryBard`](https://github.com/EvanZhouDev/bard-ai/blob/948a2aa7352102cc5e096e7def9de1c4950f2aa9/index.js#L34) and [`askAI`](https://github.com/EvanZhouDev/bard-ai/blob/948a2aa7352102cc5e096e7def9de1c4950f2aa9/index.js#L147) functions, along with [`class Chat`](https://github.com/EvanZhouDev/bard-ai/blob/948a2aa7352102cc5e096e7def9de1c4950f2aa9/index.js#L152)) in current codebase accept `SNlM0e` data as the input.
- [`queryBard`](https://github.com/EvanZhouDev/bard-ai/blob/948a2aa7352102cc5e096e7def9de1c4950f2aa9/index.js#L34) function uses `SNlM0e` data that saved in global variable of index.js file, this can be quiet tricky to manage if the Application has to use more than one bard API key, this will help upcoming users.

---
## My Changes
- Add `SNlM0e ` to `queryBard` and `askAI` functions parameter, and also to `class Chat` property.
- If `SNlM0e` is not available on `queryBard` function, it will use `SNlM0e` global data (data that saved in libary file, `index.js`)

### CODE REVIEW

**Comment by ThatXliner** [comment]
May close #11

**Comment by EvanZhouDev** [comment]
This is mostly good, but I have two questions (#1 is most important):
1. Where does the user get the `SNlM0e`? As far as I'm concerned, it gets updated on every refresh of the page, right? So you can't necessarily reuse it... (or can you?)
2. The fact that there are now two optional parameters for both `askAI`, `queryBard` and more makes it tedious to use. Any thoughts on this? Mayhaps we can just leave it like this, though.

**Comment by EvanZhouDev** [comment]
> May close #11

Not yet, because OP there requested multiple _cookies_, not necessarily multiple `SNlM0e`s

**Comment by EvanZhouDev** [comment]
Okay wait, I think I got it wrong; Is this something like what you are thinking?

```js
let SNlM0e_1 = Bard.init(COOKIE_KEY_1)
let SNlM0e_2 = Bard.init(COOKIE_KEY_2)

askAI("hi", COOKIE_KEY_1)
askAI("hi2", COOKIE_KEY_2)
```

If it is, then its all good to go, I'll just do some minor changes...

**Comment by EvanZhouDev** [comment]
As for the "two optional parameters" thing, we can possibly do two/three things
1. Do some rudimentary JS type detection (if parameter is string, then `SNlM0e`, otherwise, `useJSON`, same with second parameter)
2. If second parameter is boolean, `useJSON`. Otherwise, if Object, then have a `config` object like this:
```js
{
    useJSON: boolean,
    SNlM0e: string
}
```
3. Leave as is

**Comment by Aldhanekaa** [comment]
> As for the "two optional parameters" thing, we can possibly do two/three things
> 
> 1. Do some rudimentary JS type detection (if parameter is string, then `SNlM0e`, otherwise, `useJSON`, same with second parameter)
> 2. If second parameter is boolean, `useJSON`. Otherwise, if Object, then have a `config` object like this:
> 
> ```js
> {
>     useJSON: boolean,
>     SNlM0e: string
> }
> ```
> 
> 3. Leave as is

Ah this is a very great idea, giving flexibility!

**Comment by EvanZhouDev** [comment]
@Aldhanekaa is this correct?

> Okay wait, I think I got it wrong; Is this something like what you are thinking?
> 
> ```js
> let SNlM0e_1 = Bard.init(COOKIE_KEY_1)
> let SNlM0e_2 = Bard.init(COOKIE_KEY_2)
> 
> askAI("hi", COOKIE_KEY_1)
> askAI("hi2", COOKIE_KEY_2)
> ```
> 
> If it is, then its all good to go, I'll just do some minor changes...

**Comment by Aldhanekaa** [comment]
> As for the "two optional parameters" thing, we can possibly do two/three things
> 
> 1. Do some rudimentary JS type detection (if parameter is string, then `SNlM0e`, otherwise, `useJSON`, same with second parameter)
> 2. If second parameter is boolean, `useJSON`. Otherwise, if Object, then have a `config` object like this:
> 
> ```js
> {
>     useJSON: boolean,
>     SNlM0e: string
> }
> ```
> 
> 3. Leave as is


As the idea you mentioned above . from this following commit https://github.com/EvanZhouDev/bard-ai/pull/12/commits/4a919a8f8c2c2f39bab929f450eb1e506d5e892b , `askAI` function can be used as: 

```js
let SNlM0e_1 = Bard.init(COOKIE_KEY_1)
let SNlM0e_2 = Bard.init(COOKIE_KEY_2)

// set SNlM0e as the second argument.
askAI("hi", COOKIE_KEY_1) 
askAI("hi2", COOKIE_KEY_2)

// set useJSON as the second argument
askAI("hi bard", true) 
// the second argument, indicate useJSON as `true`, and `queryBard` function will use `SNlM0e_Global` data, which `SNlM0e` data that is being returned by `SNlM0e_2` (because it's the recent `Bard.init` call)

// set both `useJSON` and `SNlM0e` as the second argument
askAI("hi bard", {
   SNlM0e: SNlM0e_1,
   useJSON: false
}) 

```

**Comment by EvanZhouDev** [comment]
Also, I was thinking like if you do
```js
askAI("hi", COOKIE_KEY_1, true) // it will know `SNlM0e` and `useJSON` respectively
askAI("hi", true, COOKIE_KEY_1) // it will know `useJSON` and `SNlM0e` respectively
askAI("hi", true) // it will know `useJSON`
askAI("hi", COOKIE_KEY_1) // it will know `SNlM0e`
```
I can do the changes if you want?

**Comment by Aldhanekaa** [comment]
> @Aldhanekaa is this correct?
> 
> > Okay wait, I think I got it wrong; Is this something like what you are thinking?
> > ```js
> > let SNlM0e_1 = Bard.init(COOKIE_KEY_1)
> > let SNlM0e_2 = Bard.init(COOKIE_KEY_2)
> > 
> > askAI("hi", COOKIE_KEY_1)
> > askAI("hi2", COOKIE_KEY_2)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If it is, then its all good to go, I'll just do some minor changes...

Yes it can be used like you mentioned here also

**Comment by Aldhanekaa** [comment]
> Also, I was thinking like if you do
> 
> ```js
> askAI("hi", COOKIE_KEY_1, true) // it will know `SNlM0e` and `useJSON` respectively
> askAI("hi", true, COOKIE_KEY_1) // it will know `useJSON` and `SNlM0e` respectively
> askAI("hi", true) // it will know `useJSON`
> askAI("hi", COOKIE_KEY_1) // it will know `SNlM0e`
> ```
> 
> I can do the changes if you want?

Sure! You can use `rest parameter` for this! I just realised there's [`rest parameter`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters) in js 😅

**Comment by EvanZhouDev** [comment]
Yeah that's what I was thinking...

**Comment by Aldhanekaa** [comment]
> Yeah that's what I was thinking...

Yeah sorry buddy, I just realised this 😅

**Comment by EvanZhouDev** [comment]
All good; you want to make the changes, then? Or I can do it?

**Comment by Aldhanekaa** [comment]
> All good; you want to make the changes, then? Or I can do it?

I think I can make a new change now, however if you knew how to do typescript typing for `rest parameter`, I'll leave it to you because I still don't know yet how.

**Comment by EvanZhouDev** [comment]
Yeah cause I don't know either 🤣

**Comment by EvanZhouDev** [comment]
> > All good; you want to make the changes, then? Or I can do it?
> 
> I think I can make a new change now, however if you knew how to do typescript typing for `rest parameter`, I'll leave it to you because I still don't know yet how.

Go ahead, we can figure out TS later

**Comment by EvanZhouDev** [comment]
Actually, on second thought, let me try something. I'll make some edits first
I want to extract rest parameter parsing into an external function so that we can use it for both Chat and askAI, if you know what I eman

**Comment by ThatXliner** [comment]
I think it is preferred if the second parameter only accepted the (optional) config object (like most JavaScript libraries) but kept the options of accepting a Boolean indicating `useJson` for backwards compatibility.

The reasoning here is that if you’re trying to configure the API in the first place, you don’t need a “simple” API anymore.

**Comment by Aldhanekaa** [comment]
> 

Ah ok cool, no need to sorry for the changes, I was experimenting with `rest parameters` in typescript moment ago. 
Yeah, I do have Slack.

**Comment by Aldhanekaa** [comment]
> Actually, on second thought, let me try something. I'll make some edits first I want to extract rest parameter parsing into an external function so that we can use it for both Chat and askAI, if you know what I mean
> 
> EDIT: Sorry for all of the changes... do you have a Slack so we can communicate? I've added you to DMs through your public email


Ah ok cool, no need to sorry for the changes, I was experimenting with rest parameters in typescript moment ago.
Yeah, I do have Slack.

**Comment by EvanZhouDev** [comment]
I sent a invite to aldhanekadev@gmail.com

File: index.js [hunk 1]
```diff
@@ -144,18 +144,37 @@ const formatMarkdown = (text, images) => {
...
-export const askAI = async (message, useJSON = false) => {
-    if (useJSON) return await queryBard(message);
-    else return (await queryBard(message)).content;
+export const askAI = async (message, config) => {
+    let SNlM0e = undefined
+    let useJSON = false
+
+    if (typeof config == "string") {
+        SNlM0e = config
+    }else if (typeof config == "boolean") {
```
Comment (ThatXliner) [inline index.js:153]: So something like
```suggestion
    if (typeof config == "boolean") {
```

Comment (Aldhanekaa) [reply index.js]: Evan has came out with his idea; 
```
    let config = args.slice(0, 2)
    let SNlM0e = globalSNlM0e, useJSON = false;

    for (let property of config) {
        if (typeof property === "string") {
            SNlM0e = property;
        } else if (typeof property === "boolean") {
            useJSON = property;
        }
    }

```, I already implement this to my latest commit.

**Comment by EvanZhouDev** [comment]
One last quick thing, see Slack...

**Comment by EvanZhouDev** [comment]
@Aldhanekaa @ThatXliner
Please see: https://github.com/EvanZhouDev/bard-ai/issues/15
I think this solution is more efficient than everything that we have so far...
What do yall think?

**Comment by EvanZhouDev** [comment]
Hey @Aldhanekaa, I've started a Slack workspace. Feel free to discuss any of this on there, as sometimes gh issues and PRs can get a bit tedious...

**Pull request closed without merging**